joomla/joomla-cms
# Jroute between sites

Redo of #11060

Most of the work was originally done by @andrepereiradasilva. This is a redo of his work and more additions from me. 

## Summary of Changes

This PR is for allowing `JRoute` to generate URL's from admin to site and vice-versa.

This is a current problem especially if you are trying to make links from the backend to the frontend.

So, with this PR, there is no need anymore to make special code just to get the frontend link from the backend. With this PR we just need to use `JRoute`, like all other links. The same will work to links from frontend to backend. After this PR, no need to hardcode frontend to admin links.

We will be able to use following syntax:
```php
// Build a frontend URL irrespective of in which context (site/admin) the code is running
JRoute::link('site', $url, $xthml = true, $ssl = null);

// Build a backend URL irrespective of in which context (site/admin) the code is running
JRoute::link('administrator', $url, $xthml = true, $ssl = null);
```

This should not affect the existing `JRoute::_()` method.


## Testing Instructions

* Test various URLs for different extensions (articles, menu items, etc).
* You should try with the several combinations of site settings like:
    - SEF/non-SEF
    - With or without URL rewriting
    - With or without multi-language (i.e. with or without language filter plugin enabled).
    

### Test 1: Links from `admin` to `site`

* Use latest staging with sample data from at least two languages (en-GB, fr-FR)
* Add to isis template `index.php` file...
```php
$app = JFactory::getApplication();
$app->enqueueMessage('1. Admin to Admin: ' . JRoute::_('index.php'));
$app->enqueueMessage('2. Admin to Admin: ' . JRoute::_('index.php?option=com_content'));
$app->enqueueMessage('3. Admin to Site: ' . JRoute::link('site', 'index.php?option=com_content'));
$app->enqueueMessage('4. Admin to Site (MutilLanguage 1): ' . JRoute::link('site', 'index.php?option=com_content&lang=fr'));
$app->enqueueMessage('5. Admin to Site (MutilLanguage 2): ' . JRoute::link('site', 'index.php?option=com_users&lang=en'));
```

* Open any admin page. Check the URL in resulting message are correct.

### Test 2: Links from `site` to `admin`

* Add to protostar template `index.php` file...
```php
$app = JFactory::getApplication();
$app->enqueueMessage('1. Site to Site: ' . JRoute::_('index.php'));
$app->enqueueMessage('2. Site to Site: ' . JRoute::_('index.php?option=com_content'));
$app->enqueueMessage('3. Site to Admin: ' . JRoute::link('administrator', 'index.php?option=com_content'));
```

* Open any frontend page. Check the URL in resulting message are correct.


### Test 3: Code Review

Do a code review.

### Test 4: Current URLs

Check if current URL generated by JRouter are exactly the same, i.e. no change in current URL.
Probably the best is to leave the patch applied for some time, while you do other things and check there are no problems with the current URLs.

---

Thanks to @mbabker, @ggppdk and @infograf768 for helping out with this one.

Maintainers, to make sure that are no unintended consequences it would be good if this can be also reviewed from someone that understand the Router better than I do.
